Upgrade rust-lightning and expose more

a-mpch opened the pull request:
👋🏼 hey! thanks for this wasm module. I'm using it at this lightningdecoder.com [PR](https://github.com/andrerfneves/lightning-decoder/pull/94).

While working there I had an issue about SCID Paths and thought it was the library but we just for those cases using None.
But anyways, I upgraded the library to use latest rust-lightning and exposed more values that it could be used in the other PR.

Added some tests [from the vectors](https://github.com/lightning/bolts/blob/216914d492ecdf8a814bbab34fa9f6d31b394b98/bolt12/offers-test.json#L339) 

I'll be happy to add something to expose the short channels but maybe having context of why they weren't in place would be nice before updating it.

a-mpch committed:
chore: upgrade lightning and bech32

upgrade dependencies and breaking changes from both libraries.
maintains the same API for external users.

a-mpch committed:
test: add scid paths test scenario

a-mpch committed:
lib: expose more information of offer

adds quantity issuer chains and expiry to exposed offer

michael1011 commented on src/lib.rs:
```diff
@@ -52,6 +66,41 @@ impl Offer {
+    pub fn quantity(&self) -> Option<u64> {
+        match self.offer.supported_quantity() {
+            lightning::offers::offer::Quantity::Bounded(n) => Some(n.get()),
+            lightning::offers::offer::Quantity::Unbounded => Some(0),
```
Not sure if 0 is a good way to communicate unbounded. How about returning `None`?

michael1011 reviewed (changes requested):
Please fix the clippy warnings: `cargo clippy --all-targets --all-features -- -D warnings`

> I'll be happy to add something to expose the short channels but maybe having context of why they weren't in place would be nice before updating it.

We simply didn't need them

a-mpch committed:
fixup! lib: expose more information of offer

a-mpch committed:
fixup! lib: expose more information of offer

a-mpch committed:
feat: expose short channel ids alias

encode them as bytes mergin direction with the short_channel_id using lexicograpically ordered index for the direction of the channel

a-mpch commented:
@michael1011 did the changes and added a way of exposing the channels. I'm not fully convinced but it's a start. Let me know if it works from your side.

michael1011 commented on src/path.rs:
```diff
@@ -29,9 +29,35 @@ impl Hop {
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DirectedShortChannelId {
+    // the direction of the short_channel_id. 0 for the lesser node id compared
+    // lexicographically in ascending order, 1 for the greater node id.
```
```suggestion
    // The direction of the short_channel_id
    // 0 for the lesser node id compared lexicographically in ascending order, 1 for the greater node id
```

michael1011 commented on src/path.rs:
```diff
@@ -29,9 +29,35 @@ impl Hop {
+        }
+    }
+
+    pub fn encode(&self) -> Vec<u8> {
+        let mut bytes = Vec::new();
+        bytes.push(self.direction);
+        bytes.extend_from_slice(&self.short_channel_id.to_be_bytes());
+        bytes
+    }
```
Why? You cannot query channels by this format

a-mpch replied:
I'm encoding kinda like the TLV from bolt12 1 byte direction + 8 bytes channel id.
Maybe docs or do you have another suggestion? just throw the object?

michael1011 replied:
Does anyone *need* that format?

a-mpch replied:
I was planning to use it to expose the information in _almost human readable_ separating the bytes in the ui 🤷🏼

michael1011 replied:
Why encode them to bytes just to decode them again in the UI? Just expose the two fields

a-mpch replied:
sure, seems right

michael1011 commented on src/path.rs:
```diff
@@ -43,6 +69,19 @@ impl Path {
                 IntroductionNode::DirectedShortChannelId(_, _) => None,
             },
+            introduction_node_short_channel_id: match path.introduction_node() {
+                IntroductionNode::NodeId(_) => None,
```
Why ignore the node id?

a-mpch replied:
because is either of those are present not both. So we have 2 keys for the same attribute (sucks) but other way around would break the `Opt<PubKey>`.

michael1011 replied:
Either both should be exposed or neither

a-mpch replied:
sorry probably I'm not understanding  something.

We are exposing the 2 of them with different keys. Offers only have one of them encoded. That means that so some offers would have NodeId and other SCID. So we are not doing much here or ignoring. Open to explain more your thoughts? 🙏🏼

michael1011 replied:
Nevermind, it's fine

michael1011 commented on src/path.rs:
```diff
@@ -54,6 +93,19 @@ impl Path {
                 IntroductionNode::DirectedShortChannelId(_, _) => None,
             },
+            introduction_node_short_channel_id: match path.introduction_node() {
+                IntroductionNode::NodeId(_) => None,
```
Same here: why ignore the node id?

a-mpch committed:
fixup! lib: expose more information of offer

a-mpch committed:
fixup! feat: expose short channel ids alias

michael1011 commented:
Thank you!
